### PR TITLE
tests for legacy checkpoints

### LINKED ISSUE #4900: test loading legacy checkpoints
## 🚀 Feature

create persistent storage with all legacy version checkpoints and try to load and continue training

### PR DESCRIPTION
**Pull request by Borda**
## What does this PR do?

Fixes #4900

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:  

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by Borda** [comment]
OK still the need to pull the data for (or skip these envs):
- [ ] Drone with GPUs
- [ ] TPU test docker image

File: .github/workflows/ci_test-conda.yml [hunk 1]
```diff
@@ -38,6 +38,18 @@ jobs:
...
+    - name: Configure AWS credentials
+      uses: aws-actions/configure-aws-credentials@v1
+      with:
+        aws-access-key-id: ${{ secrets.AWS_ACCESS_KEY_ID }}
+        aws-secret-access-key: ${{ secrets.AWS_SECRET_KEY_ID }}
+        aws-region: us-east-1
+
+    - name: Copy files from S3
+      run: |
+        aws s3 cp --recursive s3://pl-public-data/legacy/checkpoints/ legacy/checkpoints/ #  --acl public-read
```
Comment (tchaton) [inline .github/workflows/ci_test-conda.yml:50]: Can we cache this ? Maybe use something like hash(`aws ls  s3://pl-public-data/legacy/checkpoints`)

File: .github/workflows/release-pypi.yml [hunk 2]
```diff
@@ -61,3 +61,6 @@ jobs:
       with:
         user: __token__
         password: ${{ secrets.pypi_password }}
+
+    # todo: upload checkpoint to legacy
```
Comment (tchaton) [inline .github/workflows/release-pypi.yml:65]: Great idea !

File: tests/checkpointing/test_legacy_checkpoints.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+
+@pytest.mark.parametrize("pl_version", ["1.0.0", "1.0.1", "1.0.2"])
+def test_resume_legacy_checkpoints(tmpdir, pl_version):
+    path_dir = os.path.join(LEGACY_CHECKPOINTS_PATH, pl_version)
+
+    # todo: make this as mock, so it is cleaner...
+    orig_sys_paths = list(sys.path)
+    sys.path.insert(0, path_dir)
+    from zero_training import DummyModel
```
Comment (tchaton) [inline tests/checkpointing/test_legacy_checkpoints.py:34]: Can we rename DummyModel -> LegacyCheckpointingModel

File: .github/workflows/release-pypi.yml (hunk 2 above)
Comment (Borda) [reply .github/workflows/release-pypi.yml:66]: yes, when it is done it shall be completely automatic, except adding version in pytest params

File: tests/checkpointing/test_legacy_checkpoints.py (hunk 3 above)
Comment (Borda) [reply tests/checkpointing/test_legacy_checkpoints.py:36]: well I think it is fine as it is loaded form the very same folder as saved checkpoint

File: .github/workflows/ci_test-conda.yml (hunk 1 above)
Comment (Borda) [reply .github/workflows/ci_test-conda.yml:50]: well, each model has 2KB so the total folder has about 80KB so we can add later if needed...
and then probably check your folder has all legacy checkpoint and if not pull...

Comment (Borda) [reply .github/workflows/ci_test-conda.yml:50]: but adding todo... :]

File: tests/checkpointing/test_legacy_checkpoints.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+import pytest
+
+from pytorch_lightning import Trainer
+from tests import LEGACY_PATH
+
+LEGACY_CHECKPOINTS_PATH = os.path.join(LEGACY_PATH, 'checkpoints')
+CHECKPOINT_EXTENSION = ".ckpt"
+
+
```
Comment (Borda) [inline tests/checkpointing/test_legacy_checkpoints.py:26]:
```suggestion

# todo: add more legacy checkpoints :]
```

**Review by SeanNaren — APPROVED** [review]

**Comment by Borda** [comment]
@tchaton mind review as code-owner?

**Comment by Borda** [comment]
@awaelchli @tchaton pls review ^^

**Review by awaelchli — APPROVED** [review]
looking very good! so the python script gets copied to the same location we save the checkpoint, to keep full reproducibility, right? should we maybe set a seed too? just to be sure?

**Comment by Borda** [comment]
> looking very good! so the python script gets copied to the same location we save the checkpoint, to keep full reproducibility, right? should we maybe set a seed too? just to be sure?

the setting seed is a good point even not sure if it is needed as we do not check any metrics nor weights...
we can add it follow-up PR as we need to add more checkpoints anyway, this is mainly PoC that the flow works fine :]

File: tests/checkpointing/test_legacy_checkpoints.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+# todo: add more legacy checkpoints :]
+@pytest.mark.parametrize("pl_version", [
+    "0.10.0", "1.0.0", "1.0.1", "1.0.2", "1.0.3", "1.0.4", "1.0.5", "1.0.6", "1.0.7", "1.0.8"
```
Comment (tchaton) [inline tests/checkpointing/test_legacy_checkpoints.py:29]: Any chance we could add 1.1.x too ?

**Review by tchaton — APPROVED** [review]

**Comment by tchaton** [comment]
> looking very good! so the python script gets copied to the same location we save the checkpoint, to keep full reproducibility, right? should we maybe set a seed too? just to be sure?

@Borda It seems to be copied but not used right ?

File: tests/checkpointing/test_legacy_checkpoints.py (hunk 5 above)
Comment (Borda) [reply tests/checkpointing/test_legacy_checkpoints.py:29]: yes, as mentioned https://github.com/PyTorchLightning/pytorch-lightning/pull/5223#issuecomment-752192444 this is PoC and the next PR will add `0.{6,7,8,9,10}.x` and `1.1.x` and eventually seeds...

**Comment by Borda** [comment]
> > looking very good! so the python script gets copied to the same location we save the checkpoint, to keep full reproducibility, right? should we maybe set a seed too? just to be sure?
> 
> @Borda It seems to be copied but not used right ?

yes, next PR https://github.com/PyTorchLightning/pytorch-lightning/pull/5223#issuecomment-752192444

**Review by williamFalcon — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

**Comment by Borda** [comment]
@carmocca @akihironitta wants to take it over and some more versions? :]

**Pull request merged**